### DISCUSSION TITLE
example code for ktx2 compression

### DISCUSSION THREAD

**Discussion by tom-deadline** [donmccurdy/glTF-Transform#252] [Q&A] [answered]
Hello,
I was reading docs of TextureBasisu extension and couldn't understand this part: 

> Compression is not done automatically when adding the extension as shown above — you must compress the image data first, then pass the .ktx2 payload to Texture.setImage. The glTF-Transform CLI has functions to help with this, or any similar KTX2-capable utility will work.

could you provide some example code of how it should be done?

**Comment by donmccurdy** [comment, marked as answer]
Several tools for converting textures to `.ktx2` are discussed here:

https://github.com/KhronosGroup/3D-Formats-Guidelines/blob/main/KTXArtistGuide.md

The glTF-Transform CLI is one of them, and you can find [my KTX conversion code](https://github.com/donmccurdy/glTF-Transform/blob/main/packages/cli/src/transforms/toktx.ts) in the repo as part of the CLI. Note that it relies on having KTX-Software installed to the machine, providing the `toktx` CLI program, and isn't a plain JavaScript solution.

If you need plain JavaScript and/or WASM, the basis_universal repository recently added a WASM KTX2 encoder, but I haven't worked with it yet: https://github.com/BinomialLLC/basis_universal/blob/master/webgl/ktx2_encode_test/index.html

**Reply by tom-deadline** [reply]
@donmccurdy  Thanks, a lot. Right now I can convert png to ktx2 and add textures to gltf with your TextureBasisu extension(but old ones stay there too), but how can I replace old ones with new ones. manually I would just change URI, but if I change URI with `setURI()`, I get an error from the viewer: 
![image](https://user-images.githubusercontent.com/44176198/118760970-b530cc00-b884-11eb-8ba7-74b9a6fa5075.png)


So, how can I properly replace old textures with ktx2 files?

**Reply by donmccurdy** [reply]
Once you have a KTX2 image, the code here is all that I am doing, other than attaching the extension to the document:

https://github.com/donmccurdy/glTF-Transform/blob/23ba6200ed4ca13d5e9a563365970f3b28929c77/packages/cli/src/transforms/toktx.ts#L171-L177

The version of the "glTF Validator" included in my viewer predates the `KHR_texture_basisu` extension, so I wouldn't worry too much about those validation errors. Just use `gltf-transform inspect model.glb` and confirm that the textures are all in the correct format.
